Initial docs

abcoathup opened the pull request:
Initial docs based on READMEs and CLI and Lib docs.

Reference to `import "@openzeppelin/upgrades/contracts/Initializable.sol";` will need to be updated once https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2335 merged and released.

`network-files.adoc` may need reworking.

----

Resolves #141

frangio commented on docs/modules/ROOT/pages/network-files.adoc:
```diff
@@ -0,0 +1,64 @@
+}
+----
+
+For every logic contract, besides the deployment address, the following info is also tracked:
```
I think we use "implementation" and "logic" kind of inconsistently. I also don't think either of those terms is good.

Marto has suggested we could use the term "version", and talk about "contract versions". In this sentence we would say "For every contract version, ...".

What do you think?

abcoathup replied:
Sounds good. Updated in https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/142/commits/a05349dca998d49951c5adc19c20863e5d059223

frangio commented on docs/modules/ROOT/pages/network-files.adoc:
```diff
@@ -0,0 +1,64 @@
+For every logic contract, besides the deployment address, the following info is also tracked:
+
+* `types` keeps track of all the types used in the contract or its ancestors, from basic types like `uint256` to custom `struct` types
+* `storage` tracks the storage layout of the linearized contract, referencing the types defined in the `types` section, and is used for verifying that any storage layout changes between subsequent versions are compatible
```
Here we could have a link to the FAQ entry about storage compatibility.

abcoathup replied:
Added link: https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/142/commits/c00466b4ce08e97f8149af712e1d409165297958

frangio commented on docs/modules/ROOT/pages/network-files.adoc:
```diff
@@ -0,0 +1,64 @@
+* `types` keeps track of all the types used in the contract or its ancestors, from basic types like `uint256` to custom `struct` types
+* `storage` tracks the storage layout of the linearized contract, referencing the types defined in the `types` section, and is used for verifying that any storage layout changes between subsequent versions are compatible
+
+The naming of the file will be `<network_name>.json`, but note that `<network_name>` is not taken from the name of the network's entry in the Truffle or Buidler configuration file, but is instead inferred from the canonical network id associated to the entry. 
```
Technically we use the chain id, which is not necessarily the same as the network id.

I also don't think "canonical" means anything in this context so I removed it, but if I was mistaken let me know.

```suggestion
The naming of the file will be `<network_name>.json`, but note that `<network_name>` is not taken from the name of the network's entry in the Truffle or Buidler configuration file, but is instead inferred from the chain id associated to the entry. 
```

frangio commented on docs/modules/ROOT/pages/network-files.adoc:
```diff
@@ -0,0 +1,64 @@
+[[configuration-files-in-version-control]]
+== Configuration Files in Version Control
+
+Public network files like `mainnet.json` or `ropsten.json` should be tracked in version control. These contain valuable information about your project's status in the corresponding network; the addresses of the contract implementations that have been deployed, the addresses of the proxies that have been deployed, etc. Such files should be identical for all the contributors of a project.
```
Proxy addresses are not contained in these files in the plugins.

```suggestion
Public network files like `mainnet.json` or `ropsten.json` should be tracked in version control. These contain valuable information about your project's status in the corresponding network, like the addresses of the contract implementations that have been deployed. Such files should be identical for all the contributors of a project.
```

frangio commented on docs/modules/ROOT/pages/network-files.adoc:
```diff
@@ -0,0 +1,64 @@
+
+Public network files like `mainnet.json` or `ropsten.json` should be tracked in version control. These contain valuable information about your project's status in the corresponding network; the addresses of the contract implementations that have been deployed, the addresses of the proxies that have been deployed, etc. Such files should be identical for all the contributors of a project.
+
+However, local network files like `unknown-<network_id>.json` only represent a project's deployment in a temporary local network such as `ganache-cli` that are only relevant to a single contributor of the project and should not be tracked in version control.
```
```suggestion
However, local network files like `unknown-<chain_id>.json` only represent a project's deployment in a temporary local network such as `ganache-cli` that are only relevant to a single contributor of the project and should not be tracked in version control.
```

frangio replied:
This is generally true but the "unknown" files could also correspond to other public chains that we're unaware of, like the Ethereum Classic chain. I think we should explain that.

abcoathup replied:
Added explanation for limited set of public chains
https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/142/commits/3c56793eafe3fd699e4cb6629af098df824f6684

frangio commented on docs/modules/ROOT/pages/writing-upgradeable.adoc:
```diff
@@ -0,0 +1,394 @@
+}
+----
+
+Whether using OpenZeppelin Contracts or another Ethereum Package, always make sure that the package is set up to handle upgradeable contracts.
```
```suggestion
Whether using OpenZeppelin Contracts or another smart contract library, always make sure that the package is set up to handle upgradeable contracts.
```

abcoathup replied:
Resolved in: https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/142/commits/2111130ed7681ce04989a5b15a87ed4173511454

frangio commented on docs/modules/ROOT/pages/writing-upgradeable.adoc:
```diff
@@ -0,0 +1,394 @@
+}
+----
+
+Note that it still is fine to set constants here, because the compiler https://solidity.readthedocs.io/en/latest/contracts.html#constant-state-variables[does not reserve a storage slot for these variables], and every occurrence is replaced by the respective constant expression. So the following still works with OpenZeppelin Upgrades:
```
```suggestion
Note that it is still fine to define _constant_ state variables in this way, because the compiler https://solidity.readthedocs.io/en/latest/contracts.html#constant-state-variables[does not reserve a storage slot for these variables], and every occurrence is replaced by the respective constant expression. So the following still works with OpenZeppelin Upgrades:
```

frangio commented on docs/modules/ROOT/pages/writing-upgradeable.adoc:
```diff
@@ -0,0 +1,394 @@
+}
+----
+
+The easiest way around this issue is to simply accept an instance of that contract as a parameter, and inject it after creating it:
```
```suggestion
If you would like the `ERC20` instance to be upgradeable, the easiest way to achieve that is to simply accept an instance of that contract as a parameter, and inject it after creating it:
```

frangio commented on docs/modules/ROOT/pages/writing-upgradeable.adoc:
```diff
@@ -0,0 +1,394 @@
+
+A similar effect can be achieved if the logic contract contains a `delegatecall` operation. If the contract can be made to `delegatecall` into a malicious contract that contains a `selfdestruct`, then the calling contract will be destroyed.
+
+As such, it is strongly recommended to avoid any usage of either `selfdestruct` or `delegatecall` in your contracts. If you need to include them, make absolutely sure they cannot be called by an attacker on an uninitialized logic contract.
```
I think the CLI allowed force-overriding this but we don't.

```suggestion
As such, it is not allowed to use either `selfdestruct` or `delegatecall` in your contracts.
```

abcoathup commented:
> Can you delete the API docs from the readmes, and also the FAQ, so we avoid that duplication?

I removed API & FAQ.  
Also removed How do the plugins work? (as this would have lots of links to docs FAQ) and Managing Ownership to avoid duplication.

frangio commented on README.md:
```diff
@@ -75,123 +75,6 @@ it('works before and after upgrading', async function () {
-## Usage
-
-
-## How do the plugins work?
```
I think we want to keep some content like this on the readme. The links can go to the FAQ on the docs site. But feel free to also just remove the links for now until we know what the actual links will be when we publish.

abcoathup replied:
I have added How to the plugins work and Managements back in but without links to the FAQ.

abcoathup replied:
If we know what the URL will be, I can add the links to the API and to the FAQ.

abcoathup replied:
I added links to the FAQ, and for each plugin a link to the API and the main docs.

abcoathup commented:
@frangio what will the name of the docs URL be?  Is it https://docs.openzeppelin.com/upgradesplugins or something else?

frangio commented:
According to the current config the URL will be `docs.openzeppelin.com/upgrades-plugins`.

https://github.com/OpenZeppelin/openzeppelin-upgrades/blob/8f142a4827472e12caddbb5562f2bffc86165d0e/docs/antora.yml#L1
